crates/block-producer/src/mempool/mod.rs

```diff
@@ -0,0 +1,269 @@
+#![allow(unused)]
+
+use std::{
+    collections::{btree_map::Entry, BTreeMap, BTreeSet},
+    fmt::Display,
+    ops::Sub,
+    sync::Arc,
+};
+
+use batch_graph::BatchGraph;
+use miden_objects::{
+    accounts::AccountId,
+    transaction::{ProvenTransaction, TransactionId},
+    Digest,
+};
+use miden_tx::utils::collections::KvMap;
+use transaction_graph::TransactionGraph;
+
+use crate::store::TransactionInputs;
+
+mod batch_graph;
+mod transaction_graph;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BatchJobId(u64);
+
+impl Display for BatchJobId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BatchJobId {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+pub struct BlockNumber(u32);
+
+impl Display for BlockNumber {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl BlockNumber {
+    pub fn increment(mut self) {
+        self.0 += 1;
+    }
+
+    pub fn checked_sub(&self, rhs: Self) -> Option<Self> {
+        self.0.checked_sub(rhs.0).map(Self)
+    }
+}
+
+pub struct Mempool {
+    /// The latest inflight state of each account.
+    ///
+    /// Accounts without inflight transactions are not stored.
+    account_state: BTreeMap<AccountId, (Digest, TransactionId)>,
+
+    /// The number of the next block we hand out.
+    next_inflight_block: BlockNumber,
+
+    /// Inflight transactions.
+    transactions: TransactionGraph,
+
+    /// Inflight batches.
+    batches: BatchGraph,
+
+    /// The next batches ID.
+    next_batch_id: BatchJobId,
+
+    /// Blocks which are inflight or completed but not yet considered stale.
+    block_pool: BTreeMap<BlockNumber, Vec<BatchJobId>>,
+
+    /// The current block height of the chain.
+    completed_blocks: BlockNumber,
```

Comment:
> For example, we might call `select_block` to start proving the next block even before submitting the current block to the store. This would result in effectively two inflight blocks from the perspective of the mempool.

I would probably disallow this for now as we can't do this anyway (i.e., starting construction of the next block requires that the previous block has already been computed). I think the assumption we should make is that we don't start construction of the next block until the currently constructed block is committed.

Reply:
> I would probably disallow this for now as we can't do this anyway (i.e., starting construction of the next block requires that the previous block has already been computed).

I do mean after the computation has completed, but before the block has been sent to the store. This would elide the store time.

1. select block `N` from mempool
2. compute & prove block `N`
3. select block `N+1` and begin proving
4. submit block `N` to store
5. notify mempool `N` is complete once store completes

But I'll keep it simpler for now; we must just be aware of this in the block producer then.

Reply:
Yeah - for now let's keep it simple:

1. Select block `N` from mempool.
2. Compute and prove block `N`.
3. Submit block `N` to the store.
4. Notify mempool `N` is complete once store completes.
5. Select block `N+1` from mempool
6. etc.